Add command target-set to setup tranor

fsouza opened the pull request:
This command will receive a URL and look for specific paths in this URL:
- /config.json: contains the tranor environment configuration, as
  defined in the env.go file
- /tsuru-target: contains the tsuru target to be used

The contents of these paths are downloaded and written to the proper
location: $HOME/.tranor/config.json and $HOME/.tsuru/target(s).

This change satisfies the setup process mentioned in #1.

Commit:
env: add helper function to write the configuration file

Commit:
Add command target-set

This command will receive a URL and look for specific paths in this URL:

 - /config.json: contains the tranor environment configuration, as
   defined in the env.go file
 - /tsuru-target: contains the tsuru target to be used

The contents of these paths are downloaded and written to the proper
location: $HOME/.tranor/config.json and $HOME/.tsuru/target(s).

tatiana commented:
Hi @fsouza, very good work! :)

I have two minor comments:

(1) from a tranor perspective, wouldn't it make sense for `tsuru-target` to be inside `config.json`? And, when this was processed, it would go to `./tsuru/target`?

(2) this was missed during previous code reviews, but I suggest we have in the `README.md` at least 3 things (which are currently missing):
- how to build
- how to configure
- how to run tests (if possible, with code coverage)

I've built and validated tranor with some local files, it worked smoothly. I also run the tests locally and they are all passing.

aflag commented on env.go:
```diff
@@ -59,3 +60,18 @@ func loadConfigFile() (*Config, error) {
+
+func writeConfigFile(r io.Reader) error {
+	dir := filepath.Join(os.Getenv("HOME"), ".tranor")
+	err := os.MkdirAll(dir, 0755)
```
Why MkdirAll? Do we really want to attempt to create $HOME if it does not exist?

fsouza replied:
I didn't wanted to handle the case where the directory already exists, but the side effect can be creepy, so I updated the code. Thanks!

aflag commented on main_test.go:
```diff
@@ -25,6 +35,16 @@ func TestBaseCommandsAreRegistered(t *testing.T) {
 
+func TestDefaultTargetCommandsArentRegistered(t *testing.T) {
+	manager := buildManager("tranor")
+	cmds := []string{"target-add", "target-list", "target-remove"}
```
I believe we are missing target-set.

aflag commented on main_test.go:
```diff
@@ -15,6 +15,16 @@ func TestBaseCommandsAreRegistered(t *testing.T) {
 	manager := buildManager("tranor")
 	for name, expectedCommand := range baseManager.Commands {
+		var skip bool
+		for _, c := range baseCommandsToRemove {
```
Maybe we should define our own commandsToRemove for the tests, then we can use them both here and below. If we did that, then it would be easy to see that the test's commandsToRemove was incomplete.

aflag replied:
Since target-set is not removed, we should actually use the same list of removed commands we use below, no? Then we wouldn't even need the `TestTargetSetIsRegistered` test

fsouza replied:
manager.Register fails if we reuse a command name, so we need to remove target-set before being able to register it.

We also need the `TestTargetSetIsRegistered` to ensure that we overwrite the command that comes with tsuru (the command type). I renamed the test to make it more clear.

aflag commented on main.go:
```diff
@@ -12,9 +12,15 @@ import (
+		delete(mngr.Commands, c)
+	}
 	mngr.Register(envList{})
+	mngr.Register(targetSet{})
```
How does the manager know that this is called "target-set"?

aflag commented on env_test.go:
```diff
@@ -118,3 +119,56 @@ func TestLoadConfigFileNotConfigured(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	fullPath := filepath.Join(dir, ".tranor", "config.json")
```
It seems `github.com/tsuru/tsuru/cmd` has a function called JoinWithUserDir. Any reason not to use it?

aflag commented on target.go:
```diff
@@ -0,0 +1,82 @@
+	return writeConfigFile(resp.Body)
+}
+
+func downloadTsuruTarget(server string) error {
```
Was all this code copied from tsuru? Was there any modification? Why can't we call it directly?

fsouza replied:
The piece of code used to manage targets on tsuru wasn't exported, I changed it there and updated our code. Thanks!

aflag commented on target.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+func downloadTsuruTarget(server string) error {
+	url := strings.TrimRight(server, "/") + "/tsuru-target"
```
Why does `http://tsuru-api-public-155014016.eu-west-1.elb.amazonaws.com/tsuru-target` gives me a 404 status code? Is that expected?

aflag commented:
Good job, @fsouza! It looks good, but I have a few comments -- I had more, but then I realized their answers :)

Right now we are making everything in target-set, which is fine. However, it concerns me that it does not operate well with target-add from tsuru command. After running

```
tranor target-set http://localhost:8000
```

then tsuru target-list was completely overwritten.

I wouldn't mind having the whole target-\* family in tranor and have tranor/config.json be changed depending on the chosen target (and if a target without config.json was picked, then tranor command could complain). However, having only target-set is good enough for me, as long as it does not delete the original target-list from tsuru, but merely appends to it.

Commit:
vendor: update github.com/tsuru/tsuru/cmd

Commit:
Address some comments from PR #2

 - have a single entrypoint in the remote configuration server, and
   query the target itself from config.json
 - leverage tsuru code to write and read the target
 - change code to append to target list instead of completely
   overwritting file

Commit:
travis: integrate with codecov

Commit:
readme: add codecov badge

Commit:
readme: add Go report card badge

Commit:
readme: add instructions for setup and running tests

Also included some badges, because they look nice.

Commit:
target: improve error message and code coverage

Commit:
readme: fix json syntax in the configuration example

fsouza commented:
I just updated the README with instructions for install, setup and running tests. I also added codecov integration and some badges. This should be good for merging now. Let's just wait for travis and codecov approvals.

Please let me know if you have any questions or concerns.

aflag commented on README.md:
```diff
@@ -1,5 +1,70 @@
+##Contributing and running tests
+
+Contributions are welcome! In order to run tests locally, you need to be have
+Go 1.6+ and run:
```
I think it also works on go 1.5 ( @tatiana tested it yesterday). I think it is relevant to write somewhere that it will work with 1.5 with GO15VENDOREXPERIMENT enabled. As that is the version we are currently using to compile other go projects around here.

fsouza replied:
Keep in mind that as soon as Go 1.7 is released (next week), Go 1.5 support will be dropped. I will extend the readme note and add Go 1.5.4 to travis, so we have tests running in this version as well.

aflag commented:
I liked the new changes! :) :+1: Good job!

Commit:
Add support for Go 1.5

Updating the README and travis.yml.

tatiana commented:
:+1: awesome work, @fsouza - just merged it!
